<a href="https://colab.research.google.com/github/khalil-alexander/get_forecast/blob/main/DataAnalytics2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## <font color="red"> Pip install Astropy in google colab <font/>
### <font color="Green"> We install astropy in order to convert kilometre to longitatide later <font/>

In [1]:
!pip install astropy
!pip install country_converter
!pip install shapely
!pip install cartopy
!pip install geopandas
!pip install movingpandas
!pip install hvplot
!pip install holoviews
!pip install geoviews
!pip install global_land_mask
!pip install reverse_geocode

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 47.2/47.2 kB 1.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.7/11.7 MB 81.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 78.2/78.2 kB 1.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 161.9/161.9 kB 4.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 547.3/547.3 kB 7.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 19.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.3/3.3 MB 28.7 MB/s eta 0:00:00


## <font color="red"> Install Moudles in google colab <font/>

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
from astropy import coordinates as coord
from astropy import units as u
from astropy import time
from astropy.time import Time
from astropy.coordinates import Angle

In [4]:
import pandas as pd
from bs4 import BeautifulSoup
import requests as reqs
import xml.etree.ElementTree as ET
import io
import pprint
import json
import datetime

In [5]:
import geopandas as gpd
import movingpandas as mpd
import numpy as np
from geopy.geocoders import Nominatim

In [6]:
%matplotlib inline
import matplotlib
import matplotlib.pyplot as plt
import matplotlib.ticker as mticker
import matplotlib.cm as mcm

In [7]:
import cartopy
import cartopy.crs as ccrs
import country_converter as coco
from cartopy.feature.nightshade import Nightshade
from global_land_mask import globe

In [8]:
from shapely import geometry as shpgeom
from shapely import wkt as shpwkt

In [9]:
import holoviews as hv
import hvplot.pandas
import geoviews

In [10]:
import os
from pathlib import Path
import datetime as dt
from datetime import timezone

In [11]:
from datetime import datetime
import datetime as dt
from datetime import timezone

In [12]:
import xarray as xr

In [13]:
from global_land_mask import globe
import reverse_geocode

In [14]:
def create_latlon_positions(date: str, xyz: list[float]):
    now = Time(date, scale='utc')
    cartrep = coord.CartesianRepresentation(*xyz, unit=u.km)

    gcrs = coord.GCRS(cartrep, obstime = now)
    itrs = gcrs.transform_to(coord.ITRS(obstime = now))
    loc = coord.EarthLocation(*itrs.cartesian.xyz)

    return Angle(loc.lat).deg, Angle(loc.lon).deg


In [15]:
def get_latlon(row: pd.Series):
    date = row['Date'].strftime('%Y-%m-%d %H:%M:%S')
    xyz = [row['x units(km)'], row['y units(km)'], row['z units(km)']]
    return create_latlon_positions(date, xyz)




## <font color="red"> read_xml_DateFrame <font/>
### <font color="green">A function that reads an xml file and returns the specfic position the ISS is at during different time. <font/>

<font color="blue"> We pass in the url of an XML file:https://nasa-public-data.s3.amazonaws.com/iss-coords/current/ISS_OEM/ISS.OEM_J2K_EPH.xml
<font color="blue"> We use the XML file to webscrap data on the the position of the ISS and time the ISS was in that posistion.

<font color="blue"> In order to clean up the data we use Beautifulsoup

### <font color="green">What is Beautifulsoup? <font/>

<font color="blue">Beautifulsoup is a python libary that is used to pull XML and HTML files. In this case we want to pull out an XML file. <font/>

<font color="blue">The XML file we used to extract data from: https://nasa-public-data.s3.amazonaws.com/iss-coords/current/ISS_OEM/ISS.OEM_J2K_EPH.xml <font/>

In [16]:
def read_xml_DataFrame(url):
    """
    Do web scraping from an XML file to obtain time the International Space
    Station will be at a specfic postion and the speed it is going at.

    Parameters
    ----------
    url : String
        The url of the data in XML format you want to extract from.

    Return
    ------
    df : Pandas DataFrame
        The DataFrame containing the time the International Space Station is at
        a specfic position and the speed it is traveling.
    """
    # Get the xml request
    xml = reqs.get(url)

    # Use BeautifulSoup to Organize the XML file
    soup = BeautifulSoup(xml.content,"lxml")

    # Finds all the instances 'statevector' appeared in the xml file
    state_tag = soup.find_all('statevector')

    col_name = ['Date','x units(km)',
                'y units(km)','z units(km)',
                'x_dot(km/s)',
                'y_dot(km/s)','z_dot(km/s)']
    df = pd.DataFrame(columns=col_name)

    #Loops thorugh all elements to add rows in the empty xml_df
    for element in state_tag:
        # Turns element into text and splits the '\n'
        element = element.text.split('\n')
        # Deletes First empty String
        element.pop(0)
        # Deletes Last empty Sting
        element.pop(-1)
        # Removes the 'Z' in the elements first index
        element[0] = element[0][:-1]

        # Appending list into DataFrame by adding 1 row to the DataFrame
        #df = xml_df.append(pd.DataFrame(element, columns=col_name), ignore_index=True)
        df.loc[len(df)] = element


    # Turns the "Date" collumn into a Datetime object
    df['Date'] = pd.to_datetime(df['Date'], format='%Y-%jT%H:%M:%S.%f')
    # Sets the "Date" collumn into a Datetime object
    #df.set_index("Date", inplace=True)

    # Returns the Dataframe
    return df

In [17]:
xml_url = "https://nasa-public-data.s3.amazonaws.com/iss-coords/current/ISS_OEM/ISS.OEM_J2K_EPH.xml"
%time df = read_xml_DataFrame(xml_url)

CPU times: user 10.1 s, sys: 170 ms, total: 10.3 s
Wall time: 10.8 s


In [18]:
df

Date         x units(km)          y units(km)  \
0    2025-02-12 12:00:00    1032.86303391617    -4135.01655227222   
1    2025-02-12 12:04:00  2693.5587421218202  -3359.4192497927902   
2    2025-02-12 12:08:00  4158.4600933256697  -2339.6317437399498   
3    2025-02-12 12:12:00  5321.0206661877801  -1149.7539262479099   
4    2025-02-12 12:16:00    6096.55352743612      123.73958115751   
...                  ...                 ...                  ...   
6082 2025-02-27 11:45:00  4725.6122273278197   352.30775819179598   
6083 2025-02-27 11:49:00  3839.1599846392301     1903.17044275832   
6084 2025-02-27 11:53:00    2674.73145218746   3316.2289606109498   
6085 2025-02-27 11:57:00    1316.58492360587   4489.0743177531904   
6086 2025-02-27 12:00:00    229.643996617211   5158.9603929330797   

              z units(km)          x_dot(km/s)         y_dot(km/s)  \
0        5285.24786080595     7.16296898488722  2.6374691055433002   
1      5250.1913759590498   6.5917212472323703  3.7863673738742798   
2      4832.4175386834704   5.5412630870164703    4.66003318550072   
3      4062.3141148320401   4.0874925446178603  5.1951267094584699   
4      2995.9081624519899     2.33567153768014  5.3525206292160297   
...                   ...                  ...                 ...   
6082  -4876.7302152288303  -3.0171612421375902  6.5948530993270298   
6083  -5278.7259780336399  -4.3249845490169596  6.2504577260547203   
6084  -5297.4214788776399  -5.3196592851300499  5.4534040548973604   
6085  -4931.3291171098199  -5.9294790985872803  4.2606771881374801   
6086  -4419.0464244079003  -6.1063351683023903  3.1568493905097599   

               z_dot(km/s)  
0      0.66471433707796002  
1     -0.95502205589357003  
2      -2.5051758306921501  
3      -3.8731644775054899  
4      -4.9592151147897399  
...                    ...  
6082   -2.4423904537669201  
6083  -0.88721536660181999  
6084      0.73246350063873  
6085    2.2999334681557699  
6086      3.37272993036005  

[6087 rows x 7 columns]

In [19]:
df[['lat','lon']] = df.apply(get_latlon, axis=1, result_type='expand')

In [20]:
df = df[['Date','lat','lon']]

In [21]:
df

Date        lat         lon
0    2025-02-12 12:00:00  51.324729  -38.612692
1    2025-02-12 12:04:00  50.905665  -14.883610
2    2025-02-12 12:08:00  45.665062    6.091810
3    2025-02-12 12:12:00  37.041021   22.307133
4    2025-02-12 12:16:00  26.449077   34.684131
...                  ...        ...         ...
6082 2025-02-27 11:45:00 -45.862328   30.763528
6083 2025-02-27 11:49:00 -50.981686   51.800322
6084 2025-02-27 11:53:00 -51.277893   75.480901
6085 2025-02-27 11:57:00 -46.647216   97.013198
6086 2025-02-27 12:00:00 -40.724057  110.079313

[6087 rows x 3 columns]

In [22]:
def create_latlon_positions(date: str, xyz: list[float]):
    now = Time(date, scale='utc')
    cartrep = coord.CartesianRepresentation(*xyz, unit=u.km)

    gcrs = coord.GCRS(cartrep, obstime = now)
    itrs = gcrs.transform_to(coord.ITRS(obstime = now))
    loc = coord.EarthLocation(*itrs.cartesian.xyz)

    return Angle(loc.lat).deg, Angle(loc.lon).deg


In [23]:
def get_latlon(row: pd.Series):
    date = row['Date'].strftime('%Y-%m-%d %H:%M:%S')
    xyz = [row['x units(km)'], row['y units(km)'], row['z units(km)']]
    return create_latlon_positions(date, xyz)




In [24]:
# Writing a function to return if there is land in a given latitude/longitude
# Function returns 1 if there is land and returns 0 if there is ocean.
def get_land_flag(row):
    """
    Extract the land_flag given the latitude/longitude information.
    This function provides if there is land in a given latitude/longitude. If
    not then we can assume the given latitude/longitude is on the ocean.

    Parameters
    ----------
    lat : float
       Latitude of the location
    lon : float
       Longitude of the location

    Returns
    -------
    land_flag : boolen
       If land_flag is true or not
    """
    # Gets checks if there is land in the given latitude/longitude
    land_flag = globe.is_land(row['lat'],row['lon'])
    # Returns land flag
    return land_flag

In [25]:
# Getting the name of a country with latitude/longitude
# Returns the name of the country
def get_country_name(lat: float, lon: float) -> str:
    """
    Extract the country name given the latitude/longitude information.
    This function provides a country name even when a location is on
    the ocean. We wish it was not the case.

    Parameters
    ----------
    lat : float
       Latitude of the location
    lon : float
       Longitude of the location

    Returns
    -------
    country : str
       Country name (empty string if no country)
    """
    # Turn latitude and longitude into a tuple
    lat_lon = (lat, lon),
    # Get location with geocode
    loc_name = reverse_geocode.search(lat_lon)
    # Returns Country name
    return loc_name[0].get('country', '')

In [26]:
# Using the function get_county_name we obtain the country if land_flag is true
# Returns name of country if land_flag is true, otherwise returns 'Ocean'
def obtain_country_name(row):
    """
    Extract the country name given the latitude/longitude information.
    This function provides a country name even when a location is on
    the ocean. We wish it was not the case.

    Parameters
    ----------
    row: Pandas Dataframe
       Dataframe expected to contain the latitude and longitude

       lat : float
         Latitude of the location
       lon : float
         Longitude of the location

    Returns
    -------
    country : str
       Country name (empty string if no country)
    """
    # Checks if land_flag is true
    if row['land_flag'] == 1:
        # Gets latitude and longitude
        lat = row['lat']
        lon = row['lon']
        # Gets country name
        country = get_country_name(lat, lon)
        # Retuns country name
        return country
    else:
        # Returns ocean if land_flag is false
        return "Ocean"

In [27]:
def get_overpassed_countries(df: pd.DataFrame, country_dict: dict):
    """
    Counts the amount of countrys the ISS flys over.

    Parameters
    ----------
    df : Pandas DataFrame

    Return
    ------
    count : Int
        The number of countrys the ISS flys over.
    """
    list_countries = df["Country"].unique().tolist()
    list_countries.remove("Ocean")

    for country in list_countries:
        if country in country_dict:
            country_dict[country] += 1
        else:
            country_dict[country] = 1
    return country_dict


In [28]:
def count_countrys(df: pd.DataFrame):
    """
    Counts the amount of countrys the ISS flys over.

    Parameters
    ----------
    df : Pandas DataFrame

    Return
    ------
    count : Int
        The number of countrys the ISS flys over.
    """
    country_dict = {}
    previous_name = ""
    country_count = 0
    for name, land_flag in zip(df['Country'], df['land_flag']):
        if (previous_name != name) and (name not in country_dict) and land_flag == True:
            if name not in country_dict:
                country_dict[name] = 1
                country_count += 1

        previous_name = name
    print(country_dict)
    print(country_count)
    return country_count


In [29]:
def create_list_orbits(df: pd.DataFrame) -> list:
    """

    """
    list_orbits = list()
    temp_df = df['lon']
    first_index = 0
    second_index = 0


    for i in range(len(temp_df)-1):
        if (temp_df[i] > 0) and (temp_df[i+1] < 0):
            second_index = i+1
            list_orbits.append(df[first_index:second_index])
            first_index = second_index

    list_orbits.append(df[second_index:])


    return list_orbits

In [30]:
def get_interpolated_data(df: pd.DataFrame, seconds='20s') -> pd.DataFrame:
    """

    """

    df.rename(columns={'Date': 't'}, inplace=True)

    df.set_index('t', inplace=True)
    ds = df.to_xarray()

    start_date = df.index[0] + dt.timedelta(seconds=5)
    start_date = datetime.fromtimestamp(start_date.timestamp())
    start_date = start_date.strftime('%Y-%m-%dT%H:%M:%S')

    end_date = df.index[-1] - dt.timedelta(seconds=5)
    end_date = datetime.fromtimestamp(end_date.timestamp())
    end_date = end_date.strftime('%Y-%m-%dT%H:%M:%S')

    new_dates = pd.date_range(start_date, end_date, freq=seconds)
    interpolated_data = ds.interp(t=new_dates, method='linear')
    interpolated_df = interpolated_data.to_dataframe()

    return interpolated_df

In [31]:
def interpolate_orbits(df_list: list, seconds='20s') -> list:
    count = 0
    new_df_list = list()
    for element in df_list:
        element = get_interpolated_data(element, seconds)
        element['land_flag'] = element.apply(get_land_flag, axis=1, result_type='expand')
        element['Country'] = element.apply(obtain_country_name, axis=1, result_type='expand')
        new_df_list.append(element)


    return new_df_list



In [32]:
def get_all_countrys(df_list: list) -> dict:
    """

    """
    country_dict = dict()
    for df in df_list:
        country_dict = get_overpassed_countries(df, country_dict)
    return country_dict

In [33]:
def get_all(df: pd.DataFrame, seconds='20s') -> dict:
    """
    Reads the dataframe and returns dictionary of countries the ISS flys over.
    Parameters
    ----------
    df : Pandas DataFrame

    Return
    ------
    country_dict : dict
        A dictionary of countrys the ISS flys over.
    """
    # Make a deep copy of the dataframe
    mydf = df.copy()
    # Create a list of orbits with the copy of the dataframe
    orbit_list = create_list_orbits(mydf)
    # Go through each orbit and interpolate each orbit
    interpolated_list = interpolate_orbits(orbit_list, seconds=seconds)
    # Loop through the list and add columns land_flag and Country to the element
    for element in interpolated_list:
        element['land_flag'] = element.apply(get_land_flag, axis=1, result_type='expand')
        element['Country'] = element.apply(obtain_country_name, axis=1, result_type='expand')
    print(interpolated_list[0].columns)
    country_dict = get_all_countrys(interpolated_list)
    return country_dict


In [34]:
%time test_list_15 = get_all(df, seconds='15s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 40.3 s, sys: 337 ms, total: 40.7 s
Wall time: 41.4 s


In [35]:
%time test_list_20 = get_all(df, seconds='20s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 29.6 s, sys: 87.9 ms, total: 29.7 s
Wall time: 29.8 s


In [36]:
%time test_list_25 = get_all(df, seconds='25s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 24.3 s, sys: 73 ms, total: 24.4 s
Wall time: 24.6 s


In [78]:
%time test_list_30 = get_all(df, seconds='30s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 21.6 s, sys: 83.7 ms, total: 21.6 s
Wall time: 22.2 s


In [79]:
%time test_list_10 = get_all(df, seconds='10s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 59.8 s, sys: 191 ms, total: 60 s
Wall time: 1min 4s


In [80]:
%time test_list_5 = get_all(df, seconds='5s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 1min 54s, sys: 322 ms, total: 1min 55s
Wall time: 1min 59s


In [81]:
keys15 = set(test_list_15.keys())
keys20 = set(test_list_20.keys())
keys25 = set(test_list_25.keys())
keys30 = set(test_list_30.keys())
keys10 = set(test_list_10.keys())
keys5 = set(test_list_5.keys())

In [82]:
print(len(keys15))
print(len(keys20))
print(len(keys25))
print(len(keys30))
print(len(keys10))
print(len(keys5))

169
161
161
161
171
179


In [ ]:
keys30-keys5

Create a table
First column: Freq
Second column: How long it takes to run: Computing time
Third Column: How many countries does the ISS pass: Number of Countries and

| Frequency in seconds(s) | Computing time in seconds (s) | Amount of countries ISS pass |
|-------------------------|-------------------------------|--------------------
| 5                       |             199               |        179
| 10                      |              64               |        171
| 15                      |             41.6              |        169
| 20                      |             29.8              |        161
| 25                      |             24.6              |        161
| 30                      |             22.2              |        161

In [ ]:
keys20-keys15

In [ ]:
keys15-keys10

In [ ]:
keys10-keys15

In [ ]:
keys10-keys20

In [ ]:
keys30-keys10

In [ ]:
df

In [105]:
def get_range(df: pd.DataFrame ,start_date: str, end_date: str):
    """
    Reads DataFrame, then takes the range between the start and end date.
    ----------
    df : Pandas DataFrame

    Return
    ------
    df : Pandas DataFrame
        A specific range of the df from a start date to an end date.
    """

    start_date = dt.datetime.strptime(start_date, '%Y-%m-%d %H:%M:%S')

    end_date = dt.datetime.strptime(end_date, '%Y-%m-%d %H:%M:%S')

    for element in df["Date"]:
        if element >= start_date:
            start_date = element
            break
    for element in df["Date"]:
        if element >= end_date:
            end_date = element
            break
    print(start_date)
    print(end_date)
    print(type(start_date))
    print(type(end_date))

    #start_num = df.loc[(df['Date'] >= start_date)]
    #print(start_num)
    #end_num = df.loc[(df['Date'] <= end_date)]
    #print(end_num)
    return df.loc[(df['Date'] >= start_date) & (df['Date'] <= end_date)].reset_index(drop=True)



In [113]:
start_date = dt.datetime.now()

In [114]:
start_date

datetime.datetime(2025, 2, 14, 18, 31, 16, 806679)

In [115]:
end_date = dt.datetime.now() + dt.timedelta(days=1)

In [116]:
start_date = start_date.strftime("%Y-%m-%d %H:%M:%S")

In [117]:
end_date = end_date.strftime("%Y-%m-%d %H:%M:%S")

In [118]:
day_df = get_range(df, start_date, end_date)

2025-02-14 18:32:00
2025-02-15 18:32:00
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>


test with how many countries fly over the new get_range function (use 5s freqency)

In [119]:
day_df

Date        lat         lon
0   2025-02-14 18:32:00   6.605519  -54.919056
1   2025-02-14 18:36:00  -5.609477  -46.271669
2   2025-02-14 18:40:00 -17.649783  -37.218154
3   2025-02-14 18:44:00 -29.088284  -26.768305
4   2025-02-14 18:48:00 -39.298812  -13.646288
..                  ...        ...         ...
356 2025-02-15 18:16:00 -46.092509   73.576458
357 2025-02-15 18:20:00 -37.638196   90.045671
358 2025-02-15 18:24:00 -27.142284  102.603942
359 2025-02-15 18:28:00 -15.546730  112.737200
360 2025-02-15 18:32:00  -3.431781  121.664452

[361 rows x 3 columns]

In [120]:
%time day_dict = get_all(day_df, seconds='5s')

Index(['lat', 'lon', 'land_flag', 'Country'], dtype='object')
CPU times: user 7.84 s, sys: 21.4 ms, total: 7.86 s
Wall time: 7.99 s


In [121]:
len(day_dict)

110

In [125]:
day_dict.items()

dict_items([('Suriname', 1), ('French Guiana', 1), ('Brazil', 4), ('Indonesia', 4), ('Philippines', 1), ('United States', 6), ('Mexico', 3), ('Belize', 1), ('Honduras', 2), ('Nicaragua', 2), ('Panama', 1), ('Colombia', 2), ('Peru', 2), ('Bolivia', 2), ('Paraguay', 2), ('Thailand', 1), ('Cambodia', 1), ("Lao People's Democratic Republic", 1), ('Vietnam', 2), ('China', 7), ('Korea, Republic of', 1), ('Japan', 3), ('Chile', 5), ('Argentina', 6), ('India', 3), ('Bangladesh', 1), ('Russian Federation', 7), ('South Africa', 2), ('Zimbabwe', 1), ('Mozambique', 2), ('Malawi', 2), ('Tanzania, United Republic of', 2), ('Pakistan', 2), ('Mongolia', 4), ('Falkland Islands (Malvinas)', 1), ('Angola', 2), ('Congo, The Democratic Republic of the', 2), ('South Sudan', 1), ('Sudan', 2), ('Saudi Arabia', 1), ('Kuwait', 1), ('Iraq', 2), ('Iran, Islamic Republic of', 2), ('Turkmenistan', 2), ('Uzbekistan', 2), ('Kazakhstan', 5), ('Nigeria', 1), ('Niger', 2), ('Chad', 2), ('Libyan Arab Jamahiriya', 2), ('E

create new function with country name, return how many time it will fly over that country

In [126]:
for key, value in day_dict.items():
  print(f"Key:{key}, Value:{value}")

Key:Suriname, Value:1
Key:French Guiana, Value:1
Key:Brazil, Value:4
Key:Indonesia, Value:4
Key:Philippines, Value:1
Key:United States, Value:6
Key:Mexico, Value:3
Key:Belize, Value:1
Key:Honduras, Value:2
Key:Nicaragua, Value:2
Key:Panama, Value:1
Key:Colombia, Value:2
Key:Peru, Value:2
Key:Bolivia, Value:2
Key:Paraguay, Value:2
Key:Thailand, Value:1
Key:Cambodia, Value:1
Key:Lao People's Democratic Republic, Value:1
Key:Vietnam, Value:2
Key:China, Value:7
Key:Korea, Republic of, Value:1
Key:Japan, Value:3
Key:Chile, Value:5
Key:Argentina, Value:6
Key:India, Value:3
Key:Bangladesh, Value:1
Key:Russian Federation, Value:7
Key:South Africa, Value:2
Key:Zimbabwe, Value:1
Key:Mozambique, Value:2
Key:Malawi, Value:2
Key:Tanzania, United Republic of, Value:2
Key:Pakistan, Value:2
Key:Mongolia, Value:4
Key:Falkland Islands (Malvinas), Value:1
Key:Angola, Value:2
Key:Congo, The Democratic Republic of the, Value:2
Key:South Sudan, Value:1
Key:Sudan, Value:2
Key:Saudi Arabia, Value:1
Key:Kuwait

In [127]:
def count_countryoverpass(count_dict: dict, country: str) -> int:
    """
    Counts the amount of times the ISS flies over a country.

    Parameters
    ----------
    count_dict : Dictionary

    Return
    ------
    count : Int
        The number of times the ISS passes over a country.
    """


    for key, value in count_dict.items():
        if key == country:
            return value
    print("Country not found")
    return 0

In [129]:
count_countryoverpass(day_dict, "United States")

6

given latitude and longitude return how many times the ISS will fly over it

within the radius of 20 miles of a specific point how many times will the ISS fly over.

In [ ]:
test_date = '2025-02-13 15:00:00'
test_date = dt.datetime.strptime(test_date, '%Y-%m-%d %H:%M:%S')

In [ ]:
test_date

In [ ]:
df['Date'].loc(test_date)

In [77]:
type(test_date)

datetime.datetime

In [76]:
df['Date'].info()

<class 'pandas.core.series.Series'>
Index: 6087 entries, 0 to 6086
Series name: Date
Non-Null Count  Dtype         
--------------  -----         
6087 non-null   datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 224.1 KB
